### PR TITLE
Add a fast C++ jit codepath.

### PR DESCRIPTION
**Pull request by jblespiau**
This starts a C++ jit codepath to speed up dispatch time.
It will be enabled only after we support all features and extensive
testing within Google.

Supported features:
- scalar, numpy array and DeviceArray argument support:
  - integer, floats, boolean, and complex scalars arguments are supported.
  - The jax_enable_x64 flag will be used at object-creation type to cast scalars and numpy arrays.
  - The Jax `weak_type` attribute for arguments is supported (DeviceArray and scalars).
- The donate_argnums argument.
- Use an XLA tuple for more than 100 arguments

Unsupported features:
- jax._cpp_jit on methods e.g
    @functools.partial(jax.jit, static_argnums=0)
    def _compute_log_data(self, ...)
      ...
  This is currently not supported by the C++ codepath, because "self" won't be automatically added.
- disable_jit.

This is currently only for the XLA execution path, not the tracing path.

### CODE REVIEW

File: jax/api.py [hunk 1]
```diff
@@ -354,12 +483,64 @@ def xla_computation(fun: Callable,
...
+) -> Callable:
+  """An internal implementation for `xla_computation` and `_cpp_jit`.
+
+  See `xla_computation` for the full documentation.
+
+  Args:
+    tuple_args: If `True`, the resulting XLA computation will have a single
+      tuple argument that is unpacked into the specified function arguments.
+      By default (`None`), this will be enabled when there are more than 100
+      argumments, as it will be more efficient and that XLA do not support
```
Comment (hawkinsp) [inline jax/api.py:525]: "By default (`None`), tupling will be enabled when there are more than 100 arguments, since some platforms have limits on argument arity."

File: jax/api.py [hunk 2]
```diff
@@ -354,12 +483,64 @@ def xla_computation(fun: Callable,
...
+      many arguments otherwise.
+    xla_return: If True, additional information will be returned
```
Comment (hawkinsp) [inline jax/api.py:527]: What additional information? This comment isn't really helpful: either say what information is returned or omit the comment.

What about `return_shape`?

I think commenting on both would be helpful.

File: jax/api.py [hunk 3]
```diff
@@ -354,12 +483,64 @@ def xla_computation(fun: Callable,
...
+      axis_env=axis_env,
+      in_parts=in_parts,
+      out_parts=out_parts,
+      backend=backend,
+      tuple_args=tuple_args,
+      instantiate_const_outputs=instantiate_const_outputs,
+      return_shape=return_shape)
+
+
+_XlaReturn = collections.namedtuple("_XlaReturn", [
```
Comment (hawkinsp) [inline jax/api.py:498]: Consider using `typing.NamedTuple` here and annotating types on the fields.

File: jax/api.py [hunk 4]
```diff
@@ -406,25 +597,57 @@ def computation_maker(*args, **kwargs):
     else:
       out_parts_flat = tuple(flatten_axes(
           "xla_computation out_parts", out_tree(), out_parts))
-    c = xb.make_computation_builder('xla_computation_{}'.format(fun_name))
+    c = xb.make_computation_builder("xla_computation_{}".format(fun_name))
     xla_consts = map(partial(xb.constant, c), consts)
-    xla_args = xla._xla_callable_args(c, avals, tuple_args,
+    # As we cannot modify the closure-captured `tuple_args`, we work on a copy.
```
Comment (hawkinsp) [inline jax/api.py:602]: Keep this more succinct?
```
should_tuple = tuple_args if tuple_args is not None else (len(avals) > 100)
```
and omit the comment?

File: jax/api.py [hunk 5]
```diff
@@ -406,25 +597,57 @@ def computation_maker(*args, **kwargs):
...
+                                          donated_invars, tuple_args)
+    if any(donated_invars):
+      # TODO(tomhennigan): At call time we should mark these buffers as deleted.
+      unused_donations = [
+          str(c.GetShape(a)) for a, d in zip(xla_args, donated_invars) if d
+      ]
+      warn("Some donated buffers were not usable: {}".format(
+          ", ".join(unused_donations)))
     built = c.build(out_tuple)
     if return_shape:
```
Comment (hawkinsp) [inline jax/api.py:630]: Could we remove the many choices of possible output format from the `_xla_computation` function? Suppose we always returned `_XlaReturn`, and the consumer (either `xla_computation` or `cpp_jit`) was responsible for unpacking the pieces they need.

File: tests/api_test.py [hunk 6]
```diff
@@ -41,13 +41,29 @@
...
+from jax.lib import version
 
 from jax.config import config
 config.parse_flags_with_absl()
 FLAGS = config.FLAGS
 
 
-class JitTest(jtu.JaxTestCase):
+class CPPJitTest(jtu.JaxTestCase):
+  """Shared tests between the Python and the C++ jax,jit implementations.
```
Comment (hawkinsp) [inline tests/api_test.py:52]: I might be tempted to structure this as a common test case class but to add skip decorators with TODOs to those cases that aren't supported on C++. Eventually everything will be supported everywhere and the skip decorators can go away.

File: tests/api_test.py [hunk 7]
```diff
@@ -41,13 +41,29 @@
...
+
+  Because the Python implementation supports more features, we need to have the
+  Python tests that extend the C++ tests (and not the other way around).
+  """
+
+  @staticmethod
+  def get_jit():
```
Comment (hawkinsp) [inline tests/api_test.py:59]: Couldn't we just have a `self.jit` attribute, and two subclasses that populate `jit` differently?

`self.get_jit()` is a bit wordy.

**Review by hawkinsp — CHANGES_REQUESTED** [review]
This is pretty good, but I have few changes to suggest.

File: jax/api.py (hunk 4 above)
Comment (jblespiau) [reply jax/api.py:602]: Done.

File: jax/api.py (hunk 2 above)
Comment (jblespiau) [reply jax/api.py:527]: Done.

File: tests/api_test.py (hunk 6 above)
Comment (jblespiau) [reply tests/api_test.py:52]: I am unsure I understand what you suggest.

I understand

```
Class Base:
   ...

class PythonJit(absltest.TestCase, Base)

  @property
  def jit(self):
    return _python_jit

class CPPJit(absltest.TestCase, Base)
  @unittest.SkipTest("Not supported for CPPjit")
  def test_someting(self):
```

where CPPJit overrides `test_someting` which is defined in Base.

I know I had some issue creating such base-class, and I would find it strange to duplicate all unsupported tests in CPPJit.

I am finding the current situation clear: The tests that both pass are in CPPJit.
The test that only the Python code pass, are in Pythonjit.

File: jax/api.py [hunk 8]
```diff
@@ -354,12 +482,72 @@ def xla_computation(fun: Callable,
...
+    fun: Callable,
+    static_argnums: Union[int, Iterable[int]] = (),
+    axis_env: Optional[Sequence[Tuple[AxisName, int]]] = None,
+    in_parts=None,
+    out_parts=None,
+    backend: Optional[str] = None,
+    tuple_args: Optional[bool] = None,
+    instantiate_const_outputs: Optional[bool] = True,
+    return_shape: bool = False,
+    xla_return: bool = False,
```
Comment (hawkinsp) [inline jax/api.py:525]: Is this parameter now stale? I don't think it's ever set?

File: jax/api.py [hunk 9]
```diff
@@ -354,12 +482,72 @@ def xla_computation(fun: Callable,
...
+  """An internal implementation for `xla_computation` and `_cpp_jit`.
+
+  See `xla_computation` for the full documentation.
+
+  Args:
+    tuple_args: By default (None), tupling will be enabled when there are more
+      than 100 arguments, since some platforms have limits on argument arity.
+
+  Returns:
+    An `_XlaReturn` object, some field possibly being None depending on
```
Comment (hawkinsp) [inline jax/api.py:537]: The comment also seems stale?

File: tests/api_test.py (hunk 6 above)
Comment (hawkinsp) [reply tests/api_test.py:52]: I guess it's fine as is.

**Review by hawkinsp — CHANGES_REQUESTED** [review]

File: jax/api.py (hunk 8 above)
Comment (jblespiau) [reply jax/api.py:525]: Done.

File: jax/api.py (hunk 9 above)
Comment (jblespiau) [reply jax/api.py:537]: Yes. I was waiting for the anwser to the other one, to know which action to take, before updating everything.

it's also because I'm in Google3, on Fig, so each time to make my changes visible, I need to Amend + Upload + run the export script. So I had some changes but not all.

File: tests/api_test.py [hunk 10]
```diff
@@ -41,13 +41,42 @@
...
+                           lambda: self.jit(3))
+
+  def test_jit_of_generator(self):
+
+    def gen(x):
+      yield x
+
+    self.assertRaisesRegex(TypeError,
+                           "Expected a function, got a generator function.*",
+                           lambda: self.jit(gen))
```
Comment (jblespiau) [inline tests/api_test.py:79]: I took advance of this CL, to fix these 2 tests also^^

**Review by hawkinsp — APPROVED** [review]
I'll merge when I get a green presubmit.
